docs_src/_autosummary/linuxmusterTools.lmnconfig.samba.rst
linuxmusterTools.lmnconfig.samba
================================

.. automodule:: linuxmusterTools.lmnconfig.samba

   
   
   

   
   
   .. rubric:: Functions

   .. autosummary::
   
      parse_log_level
   
   

   
   
   

   
   
   





docs_src/_autosummary/linuxmusterTools.lmnconfig.setup.rst
linuxmusterTools.lmnconfig.setup
================================

.. automodule:: linuxmusterTools.lmnconfig.setup

   
   
   

   
   
   

   
   
   .. rubric:: Classes

   .. autosummary::
   
      SetupConfig
   
   

   
   
   





debian/changelog
linuxmuster-tools7 (7.2.47) lmn72; urgency=medium

  * Add documentation
  * Add examples
  * Rewrite logger
  * [ldapwriter] Add GroupWriter
  * [ldapwriter] Fix resetting thumbnailPhoto
  * [ldapwriter] Fix quota types 

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Mon, 20 Jan 2025 12:29:04 +0100

linuxmuster-tools7 (7.2.46) lmn72-testing; urgency=medium

  * Use env var instead of option

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Mon, 30 Dec 2024 19:45:00 +0100

linuxmuster-tools7 (7.2.45) lmn72-testing; urgency=medium

  * Temporary fix for newer pip versions

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Mon, 30 Dec 2024 17:15:48 +0100

linuxmuster-tools7 (7.2.44) lmn72; urgency=medium

  * [common] Convert sophomorix time as human readable
  * [common] Check user's status in attic 
  * [ldapreader] Get all rooms and a specific room
  * [ldapreader] Fix subtree search in groups requests
  * [ldapwriter] Check devices hostnames
  * [ldapwriter] Add OU and CN in Ldap
  * [lmnconfig] Add samba, sophomorix and setup config
  * [lmnconfig] Add LDAP_CONTEXT and SAMBA_TLD

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Mon, 09 Dec 2024 16:46:40 +0100

linuxmuster-tools7 (7.2.43) lmn72; urgency=medium

  * [ldapreader] Fix some school specific requests
  * [ldapwriter] Remove old write functions
  * [ldapwriter] Move entry to new OU/DN
  * [ldapwriter] Add Objectwriter
  * [ldapwriter] Add / Remove member from subtree
  * [ldapwriter] Move device to new room
  * [ldapwriter] Rename device 
  * [samba] Read smb.conf with right delimiter (thanks @hermanntoast)

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Mon, 18 Nov 2024 18:10:27 +0100

linuxmuster-tools7 (7.2.42) lmn72; urgency=medium

  * [ldapconnector] Introduce new writers
  * [ldapconnector] Optimize requests time 

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Sun, 03 Nov 2024 11:00:21 +0100

linuxmuster-tools7 (7.2.41) lmn72; urgency=medium

  * [ldapreader] Refactor with common parent methods
  * [ldapreader] Basic exports as json and csv
  * [quotas] Check quotas for DFS shares 

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Mon, 07 Oct 2024 12:27:58 +0200

linuxmuster-tools7 (7.2.40) lmn72; urgency=medium

  * [ldapreader] Fix printer search
  * [samba] Load DFS shares 

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Mon, 30 Sep 2024 21:55:30 +0200

linuxmuster-tools7 (7.2.39) lmn72; urgency=medium

  * [ldapreader] No customFields object for Webui import. 

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Sun, 29 Sep 2024 14:09:02 +0200

linuxmuster-tools7 (7.2.38) lmn72; urgency=medium

  * [ldapreader] Ignore SCHOOLMARKER for globaladmins. 
  * [ldapwriter] Fix delete attribute.

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Sat, 28 Sep 2024 12:59:32 +0200

linuxmuster-tools7 (7.2.37) lmn72; urgency=medium

  * [ldapreader] Add dn filter and refactor url decorators
  * [ldapreader] Custom fields object as user attribute
  * [quotas] Better handle quotas errors 

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Sat, 28 Sep 2024 10:42:30 +0200

linuxmuster-tools7 (7.2.36) lmn72; urgency=medium

  * [ldapreader] Get object from dn.
  * [ldapreader] Add common object model.
  * [ldapreader] Real ldap_reader and ldapconnector objects.
  * [ldapwriter] Add ldap entry, group or OU.
  * [ldapwriter] Delete entry.
  * [ldapwriter] Separate del and set methods for each model.

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Thu, 19 Sep 2024 18:28:33 +0200

linuxmuster-tools7 (7.2.35) lmn72; urgency=medium

  * [ldapreader] Fix conflicts for parallel requests
  * [ldapreader] Missing members in projects from sophomorix output

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Sat, 07 Sep 2024 17:24:18 +0200

linuxmuster-tools7 (7.2.34) lmn72; urgency=medium

  * [ldapreader] asdict for user model
  * [ldapreader] Handle subdn for global search
  * [ldapreader] Add printer model and get all members
  * [ldapreader] Get from dn in router 
  * [ldapwriter] Setting list values

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Thu, 05 Sep 2024 21:16:07 +0200

linuxmuster-tools7 (7.2.33) lmn72; urgency=medium

  * [ldapreader] Add proxyAddresses und IntrinsicMulti1 to groups and projects.
  * [ldapreader] Differenciate AD group.
  * [ldapreader] Search for displayName too.
  * [ldapreader] Add management group to ldap_writer.

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Wed, 24 Jul 2024 17:29:05 +0200

linuxmuster-tools7 (7.2.32) lmn72; urgency=medium

  * [ldapreader] Add displayName to LDAP objects 

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Mon, 22 Jul 2024 18:49:44 +0200

linuxmuster-tools7 (7.2.31) lmn72; urgency=medium

  * [samba] Scan gz logs too 

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Mon, 22 Jul 2024 18:03:25 +0200

linuxmuster-tools7 (7.2.30) lmn72; urgency=medium

  * [ldapreader] Add attributes to search answers 

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Mon, 22 Jul 2024 17:21:45 +0200

linuxmuster-tools7 (7.2.29) lmn72; urgency=medium

  * [samba] Sort logs by reverse date 

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Thu, 18 Jul 2024 18:21:58 +0200

linuxmuster-tools7 (7.2.28) lmn72; urgency=medium

  * [samba] Handle missing log level parameter 

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Thu, 18 Jul 2024 18:16:14 +0200

linuxmuster-tools7 (7.2.27) lmn72; urgency=medium

  * [samba] Get last login 

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Thu, 18 Jul 2024 16:06:24 +0200

linuxmuster-tools7 (7.2.26) lmn72; urgency=medium

  * [ldapreader] Getval for collections
  * [ldapreader] Difference between units and sophomorix-group
  * [ldapreader] Get projects with special chars 

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Tue, 09 Jul 2024 12:29:09 +0200

linuxmuster-tools7 (7.2.25) lmn72; urgency=medium

  * [validator] Better terminology. 

 -- Arnaud Kientz <arnaud@linuxmuster.net>  Mon, 13 May 2024 09:39:09 +0200

linuxmuster-tools7 (7.2.24) lmn72; urgency=medium

  * [ldapreader] Check if linuxmuster is configured. 


usr/lib/python3/dist-packages/linuxmusterTools/linbo/config.py
from dataclasses import dataclass, field
from datetime import datetime
import os
import locale
import time

from ..devices import Devices


@dataclass
class Partition:
    Bootable: bool
    Dev: str
    FSType: str
    Id: int
    Label: str
    Size: str

@dataclass
class OS:
    Append: str
    Autostart: bool
    AutostartTimeout: int
    BaseImage: str
    Boot: str
    DefaultAction: str
    Description: str
    Hidden: bool
    IconName: str
    Initrd: str
    Kernel: str
    NewEnabled: bool
    Root: str
    StartEnabled: bool
    SyncEnabled: bool
    Version: str

@dataclass
class Linbo:
    AutoFormat: bool
    AutoInitCache: bool
    AutoPartition: bool
    Cache: str
    DownloadType: str
    Group: str
    GuiDisabled: bool
    KernelOptions: str
    Locale: str
    RootTimeout: int
    Server: str
    SystemType: str
    UseMinimalLayout: bool

@dataclass
class LinboConfig:
    LINBO: Linbo
    Partitions: list
    OS: list

class LinboConfigManager:
    pass

## The following functions need to be rewritten

LINBO_PATH = '/srv/linbo'

def read_config(group):
    """
    Get the os config from linbo config file start.conf.<group>

    :param group: Linbo group
    :type group: string
    :return: Config as list of dict
    :rtype: list of dict
    """

    path = os.path.join(LINBO_PATH, 'start.conf.'+group)
    osConfig = []
    if os.path.isfile(path):
        for line in open(path):
            line = line.split('#')[0].strip()

            if line.startswith('['):
                section = {}
                section_name = line.strip('[]')
                if section_name == 'OS':
                    osConfig.append(section)
            elif '=' in line:
                k, v = line.split('=', 1)
                v = v.strip()
                if v in ['yes', 'no']:
                    v = v == 'yes'
                section[k.strip()] = v
        return osConfig
    return None

def last_sync(workstation, image):
    """
    Get the date of the last sync date for a workstation w.

    :param w: Workstation
    :type w: string
    :param image: Name of the image file
    :type image: string
    :return: Last synchronisation time
    :rtype: datetime
    """

    statusfile = f'/var/log/linuxmuster/linbo/{workstation}_image.status'
    image_last_sync, diff_last_sync = '0','0'
    diff_image = image.replace('.qcow2', '.qdiff')

    if os.path.isfile(statusfile) and os.stat(statusfile).st_size != 0:
        for line in open(statusfile, 'r').readlines():
            if image in line:
                image_last_sync = line.rstrip().split(' ')[0]
            if diff_image in line:
                diff_last_sync = line.strip().split(' ')[0]

    last = max(image_last_sync, diff_last_sync)

    if last == '0':
        return False

    ## Linbo locale is en_GB, not necessarily the server locale
    saved = locale.setlocale(locale.LC_ALL)
    locale.setlocale(locale.LC_ALL, 'C.UTF-8')
    last = datetime.strptime(last, '%Y%m%d%H%M')
    locale.setlocale(locale.LC_ALL, saved)

    last = time.mktime(last.timetuple())
    return last

def group_os(workstations):
    """
    Get all os infos from linbo config file and inject it in workstations dict.
    The workstations dict is set in the function list_workstations().

    :param workstations: Dict containing all workstations
    :type workstations: dict
    :return: Completed workstations dict with linbo informations
    :rtype: dict
    """

    for group in workstations.keys():
        workstations[group]['os'] = []
        config = read_config(group)
        if config is not None:
            workstations[group]['power'] = {
                'run_halt': 0,
                'timeout': 1
                }
            workstations[group]['auto'] = {
                'disable_gui': 0,
                'bypass': 0,
                'wol': 0,
                'prestart': 0,
                'partition': 0,
            }
            for osConfig in config:
                if osConfig['SyncEnabled'] or osConfig['NewEnabled']:
                    tmpDict = {
                                'baseimage': osConfig['BaseImage'],
                                'partition': osConfig['Root'][-1],
                                'new_enabled': osConfig['NewEnabled'],
                                'start_enabled': osConfig['StartEnabled'],
                                'run_format': 0,
                                'run_sync':0,
                                'run_start':0,
                        }
                    workstations[group]['os'].append(tmpDict)

    return workstations

def list_workstations(school='default-school', groups=[]):
    """
    Generate a dict with workstations and parameters out of devices file

    :param context: user context set in views.py
    :return: Dict with all linbo informations for all workstations.
    :rtype: dict
    """

    devices_dict = {}
    devices_manager = Devices(school=school)
    devices = devices_manager.filter(groups=groups)

    for device in devices:
        if school != 'default-school':
            if device['hostname']:
                device['hostname'] = f'{school}-{device["hostname"]}'
        if os.path.isfile(os.path.join(LINBO_PATH, 'start.conf.'+str(device['group']))):
            if device['pxeFlag'] != '1' and device['pxeFlag'] != "2":
                continue
            elif device['group'] not in devices_dict.keys():
                devices_dict[device['group']] = {'grp': device['group'], 'hosts': [device]}
            else:
                devices_dict[device['group']]['hosts'].append(device)
    return group_os(devices_dict)

def last_sync_all(workstations):
    """
    Add last synchronisation informations into the workstations dict,
    and status attribute to use as class for bootstrap.

    :param workstations: Dict of workstations set in list_workstations().
    :type workstations: dict
    :return: Completed dict of workstations
    :rtype: dict
    """

    today = time.mktime(datetime.now().timetuple())

    for group, grpDict in sorted(workstations.items()):
            for host in grpDict['hosts']:
                host['images'] = []
                host['sync'] = {}
                for image in workstations[group]['os']:
                    last = last_sync(host['hostname'], image['baseimage'])
                    date = last if last else "Never"
                    tmpDict = {
                            'date': date,
                    }
                    if date == "Never" or (today - date > 30*24*3600):
                        tmpDict['status'] = "danger"
                    elif today - date > 7*24*3600:
                        tmpDict['status'] = "warning"
                    else:
                        tmpDict['status'] = "success"
                    host['sync'][image['baseimage']] = tmpDict
                    host['images'].append(image['baseimage'])